### ISSUE TITLE
Correctly determining parent page

### ISSUE THREAD

**Issue by ksachdeva** [rotorgames/Rg.Plugins.Popup#49]
Dear Kirill,

First of all thanks for a wonderful library. 

I have a bit complex application where there are multiple navigation stacks and facing an issue when using Rx.Plugins.Popup in few situations.

After reading your code, I believe the issue is coming from the implementation of 

```
private static Page GetParentPage()
        {
            Page lastPage = Application.Current.MainPage.Navigation.ModalStack.LastOrDefault();

            if (lastPage == null)
                lastPage = Application.Current.MainPage;

            return lastPage;
        }
```

Here you are using the Navigation of MainPage, now a relatively complex application could have a page that has its own Navigation Stack. I wanted to show the popup on the page that has its own navigation stack but it appears on the stack of MainPage (as per your implementation)

I believe that this could be fixed by letting the caller of your API provide the parent page on which you should show the Popup. If the caller does not provide the parent page then you use your GetParentPage implementation. Have not tested this suggestion myself but in theory should work.

Again, thanks for the great work you have done on this library.

Regards
Kapil

**Comment by rotorgames** [comment]
@ksachdeva Hi friend. Can you show your custom navigation to me?

**Comment by ksachdeva** [comment]
yes. let me create a smaller project to reproduce the issue.

**Comment by rotorgames** [comment]
@ksachdeva Ok.

**Comment by ksachdeva** [comment]
Here you go :

I just took your demo application and added the use case to it -

[pages.zip](https://github.com/rotorgames/Rg.Plugins.Popup/files/527896/pages.zip)

Here is a sequence of pages:

MainPage ---> BModalPage ---> AModalPage

MainPage has a button that opens BModalPage using Navigation.PushModalAsync(page);

BModalPage has a button 'Add Modal' that shows AModalPage however this is where things are interesting/different. Instead of using doing PushModalAsync here is what I do 

```
private async void OnAddModalPage(object sender, EventArgs e)
        {
            AModalPage aModalPage = new AModalPage();

            NavigationPage navPage = new NavigationPage(aModalPage);

            await this.Navigation.PushModalAsync(navPage);
        }
```

On the AModalPage I have buttons 'Show Popup' and 'Close'.  Clicking on 'Show Popup' will not show anything on AModalPage. Click on 'Close' will take you back to 'BModalPage' and you will see the popup there.

You should be able to simply unzip the attached zip in Demo application.

What has happened in my sample application is that I have changed the NavigationPage. Your GetParent() method takes the NavigationPage of the Root of the application and hence the issue.

I will not try to modify your code to see if I can get it to work with out breaking backward compatibility.

Regards
Kapil

**Comment by ksachdeva** [comment]
Here is how I made it work on iOS :

PopupNavigationIOS.cs

```
internal class PopupNavigationIOS : IPopupNavigation
    {
        public void AddPopup(PopupPage page)
        {
            var renderer = page.GetOrCreateRenderer();

            // KS : passing parent as argument
            GetTopViewController(page.Parent as Page).View.AddSubview(renderer.NativeView);
        }

        public void RemovePopup(PopupPage page)
        {
            var renderer = page.GetOrCreateRenderer();
            var viewController = renderer?.ViewController;

            if (viewController != null && !viewController.IsBeingDismissed)
                renderer.NativeView.RemoveFromSuperview();

        }

        private UIViewController GetTopViewController(Page parent)
        {
            // KS : if parent is not null (Can it be not null ??) then 
            if (parent != null)
                return parent.GetOrCreateRenderer().ViewController;

            var navigation = Application.Current.MainPage.Navigation;

            if (navigation.ModalStack.Count > 0)
                return navigation.ModalStack.First().GetOrCreateRenderer().ViewController;
            
            return UIApplication.SharedApplication.KeyWindow.RootViewController;
        }
    }
```

In PopupNavigation.cs

```

public static Task PushAsync(PopupPage page, bool animate = true, Page parentPage = null)
        {
            var task = new TaskCompletionSource<bool>();
            var parent = parentPage ?? GetParentPage();
            page.Parent = parent;
            if (animate)
            {
                page.PreparingAnimation();
                page.ExecuteWhenAppearingOnce(async () =>
                {
                    await page.AppearingAnimation();
                    task.TrySetResult(true);
                });
            }
            DependencyService.Get<IPopupNavigation>().AddPopup(page);
            _popupStack.Add(page);
            if(!animate) task.TrySetResult(true);
            return task.Task;
        }
```

now going to try on Android

**Comment by ksachdeva** [comment]
The use case is working on Android with out my modifications !

**Comment by rotorgames** [comment]
@ksachdeva Add parentPage argument is bad practice. I can't change architecture for fix one bug.

I must find other solution.

**Comment by ksachdeva** [comment]
Could you explain why passing the parentPage would be a bad practice ? 

Regards & thanks
Kapil

**Comment by rotorgames** [comment]
@ksachdeva It is bad practice because I must change signature of method but I don't want to do it.

Why do You don't use `INavigation` for your navigation page and don't use MainPage for it?

**Comment by rotorgames** [comment]
@ksachdeva Hi friend. I understood your problem. I will fix it. Please wait.

**Comment by rotorgames** [comment]
@ksachdeva P.S. If I post `.nupkg` file to you. You can test it?

**Comment by rotorgames** [comment]
@ksachdeva I fixed this problem. Please, pull last commits from master and try it.

**Comment by ksachdeva** [comment]
@rotorgames Hi Friend, you are the best. I liked your approach. Just tested it on Android & iOS and it works. 

Thanks again.

**Comment by rotorgames** [comment]
@ksachdeva Ok. I can post `.nupkg` file to your email if need. New version in nuget will not be soon.

**Comment by ksachdeva** [comment]
Ok.

**Comment by rotorgames** [comment]
@ksachdeva Give me your email please.

**Comment by ksachdeva** [comment]
Have sent it to you by a separate email. Thanks.

**Comment by rotorgames** [comment]
@ksachdeva Sended.

**Comment by rotorgames** [comment]
Fixed in `v1.0.4`
